### PR TITLE
FEniCS: update of packages to support 2019.1.0

### PR DESCRIPTION
**Pull request by emai-imcs**
PR to introduce FEniCS 2019.1.0 support for spack, which is the latest stable version prior to FEniCS-X. The PR is based on the previous PR by [benzwick](https://github.com/spack/spack/pull/10240) and the local spack repo of [JaroslavHron](https://github.com/JaroslavHron/spack_repo/tree/master/fenics).

The changes comprise the 4 python packages,

- FFC: py-fenics-ffc,
- UFL: py-fenics-ufl,
- FIAT: py-fenics-fiat,
- Dijitso: py-fenics-dijitso,

and the main Dolfin package,

- fenics2019.

The dolfin package builds and installs the c++ as well as the python interface of FEniCS. To fix an compilation error with the latest boost package a patch file is provided.

### CODE REVIEW

File: var/spack/repos/builtin/packages/fenics2019/package.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    depends_on('qt', when='+qt')
+
+    depends_on('py-pybind11', type=('build', 'run'))
+    depends_on('cmake@3.17.3:', type='build')
+
+    depends_on('py-setuptools', type='build', when='+python')
+    depends_on('py-pkgconfig', type='run', when='+python')
+    depends_on('py-sphinx@1.0.1:', when='+doc', type='build')
+
+    def cmake_is_on(self, option):
```
Comment (haampie) [inline var/spack/repos/builtin/packages/fenics2019/package.py]: Could you replace this with the builtin `self.define_from_variant`?

https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake

File: var/spack/repos/builtin/packages/fenics2019/package.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                self.cmake_is_on('py-slepc4py')),
+            '-DDOLFIN_ENABLE_SPHINX:BOOL={0}'.format(self.cmake_is_on('py-sphinx')),
+            '-DDOLFIN_ENABLE_TRILINOS:BOOL={0}'.format(
+                self.cmake_is_on('trilinos')),
+            '-DDOLFIN_ENABLE_UMFPACK:BOOL={0}'.format(
+                self.cmake_is_on('suite-sparse')),
+            '-DDOLFIN_ENABLE_VTK:BOOL={0}'.format(self.cmake_is_on('vtk')),
+            '-DDOLFIN_ENABLE_ZLIB:BOOL={0}'.format(self.cmake_is_on('zlib')),
+        ]
+        os.environ['PETSC_DIR'] = spec['petsc'].prefix
```
Comment (haampie) [inline var/spack/repos/builtin/packages/fenics2019/package.py]: Is this necessary, or can these be found through CMAKE_PREFIX_PATH too, which spack sets by default for deps?

File: var/spack/repos/builtin/packages/fenics2019/package.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        env.set('DOLFIN_DIR', self.prefix)
+
+    # build python interface of dolfin
+    @run_after('install')
+    def install_python_interface(self):
+        if '+python' in self.spec:
+            cd('python')
+            python('setup.py', 'install',
+                   '--single-version-externally-managed',
+                   '--root=/', '--prefix={0}'.format(self.prefix))
```
Comment (haampie) [inline var/spack/repos/builtin/packages/fenics2019/package.py]: I don't know much about python, but just to be sure: what does `--root=/` do?

File: var/spack/repos/builtin/packages/py-fenics-fiat/package.py [hunk 4]
```diff
@@ -23,10 +23,6 @@ class PyFenicsFiat(PythonPackage):
 
     version("master", branch="master")
     version('2019.1.0',       sha256='2a6d175a825ed725843918ef28846edbcf710a879c2fe8caaeda77b1ce9b9a1c')
```
Comment (haampie) [inline var/spack/repos/builtin/packages/py-fenics-fiat/package.py]: Why are these versions deleted?

**Review by haampie — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/fenics2019/package.py (hunk 1 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics2019/package.py]: agreed, will be changed in next commit

File: var/spack/repos/builtin/packages/fenics2019/package.py (hunk 2 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics2019/package.py]: agreed, set of os.environ will be removed

File: var/spack/repos/builtin/packages/fenics2019/package.py (hunk 3 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics2019/package.py]: The `--root=/` option is required by the `--single-version-externally-managed` flag and installs everything to the alternate root directory as specified. I found this option in the package from [JaroslavHron](https://github.com/JaroslavHron/spack_repo/tree/master/fenics/packages/fenics), and it provides a working way to get the dolfin python interface installed. Note that by only using the `--prefix` I did not manage to successfully import the dolfin package in python.

However I suggest to replace it with a simple `--user` flag, where the python packages are installed in the users site-package path and which I have successfully tested on my machine as well. 
`python('setup.py', 'install', '--user')`

File: var/spack/repos/builtin/packages/py-fenics-fiat/package.py (hunk 4 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-fiat/package.py]: The intend was to support the 2019.1.0 version of fenics only, hence to delete the old versions within the dependent python packages (ffc, ufl, fiat, dijitso). However I agree it might be easier to keep the ufl, fiat and dijitso package files as they are and to refer to the required version in the fenics2019 and ffc package.

Comment (mayrmt) [reply var/spack/repos/builtin/packages/py-fenics-fiat/package.py]: @emai-imcs Since `spack` is about providing multipler versions and allow each user to pick its favorite at install time, it's actually intended to keep the other versions around.

File: var/spack/repos/builtin/packages/fenics2019/package.py (hunk 3 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/fenics2019/package.py]: Pretty sure @adamjstewart has a more informed opinion about this

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/fenics2019/package.py]: The current logic is correct, we want to install to the package installation prefix, not a shared site-packages path. The only change I would suggest is to replace `python('setup.py', 'install'` with `setup_py('install'`, which adds additional logic to ignore user configs, making more reproducible builds.

File: var/spack/repos/builtin/packages/py-fenics-fiat/package.py (hunk 4 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-fiat/package.py]: See #20767 and #15896 for the deprecation model I would like to move towards

**Comment by adamjstewart** [comment]
We should update the existing `fenics` package instead of creating a new `fenics2019` package.

**Comment by emai-imcs** [comment]
The existing `fenics` package is based on version `2016.1.0` where I can't say how useful the old versions are. According to the [FEniCS Roadmap](https://fenicsproject.org/fenics-project-roadmap-2019/) I expect the 2019 version as the final one prior the shift to FEniCS-X.

Please note that from version 2018.1.0 fenics switched from instant to the dijitso for just-in-time compiling which needs to be addressed in package file if backward compatibility shall be provided.

**Comment by emai-imcs** [comment]
Following the comment from  @adamjstewart the existing `fenics` package is extended to include the 2019.1.0 version. In addition the versions for 2018.1.0, 2017.2.0 and 2016.2.0 were added.

The amended `fenics` package shall not be considered as final, but as a starting point for further work to be done.

1. While I could successfully install fenics 2019.1.0 and 2018.1.0, I haven't tested installation of the older versions. Due to 
increasing compatibility issues, as for example with the recent PETSc version, I would assume further patches to be necessary. Maybe the reviewers can advise if it would be acceptable to address this at a later point.

2. An issue with the `py-dijitso` installation is observed, where no resolution is found yet. Following a dolfin import in python, `form dolfin import *`, calling up dijitso by `Expression("1",degree=0)` raises an error that dijitso is unable to compile the C++ code as it fails to load an existing file. Any advise on what might cause the issue is welcome.

File: var/spack/repos/builtin/packages/fenics2019/package.py (hunk 3 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics2019/package.py]: Suggest changes are reflected in latest commit.

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 5]
```diff
@@ -10,26 +10,37 @@ class Fenics(CMakePackage):
...
+
+    variant('python',       default=True,  description='Compile with Python interface')
     variant('hdf5',         default=True,  description='Compile with HDF5')
     variant('parmetis',     default=True,  description='Compile with ParMETIS')
     variant('scotch',       default=True,  description='Compile with Scotch')
     variant('petsc',        default=True,  description='Compile with PETSc')
     variant('slepc',        default=True,  description='Compile with SLEPc')
-    variant('trilinos',     default=True,  description='Compile with Trilinos')
+    variant('py-petsc4py',  default=True,  description='Use PETSC4py')
+    variant('py-slepc4py',  default=True,  description='Use SLEPc4py')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]: I think it would be more clear if these variants were called `petsc4py` and `slepc4py`, without the `py-`. Also, is it possible to enable these variants when `~python`, or do we need to add a conflict for that?

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 6]
```diff
@@ -43,135 +54,99 @@ class Fenics(CMakePackage):
...
+        depends_on('py-fenics-ufl{0}'.format(wver), type=('build', 'run'), when=wver)
+        if ver in ['2019.1.0', '2017.2.0']:
+            wver = wver + '.post0'
+        depends_on('py-fenics-ffc{0}'.format(wver), type=('build', 'run'), when=wver)
+
+    # package dependencies
+    depends_on('python@3:', type=('build', 'run'), when='+python')
     depends_on('eigen@3.2.0:')
-    depends_on('boost+filesystem+program_options+system+iostreams+timer+regex+chrono')
+    depends_on('pkg-config')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]:
```suggestion
    depends_on('pkgconfig')
```
Spack has a virtual package for this

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 7]
```diff
@@ -43,135 +54,99 @@ class Fenics(CMakePackage):
...
 
+    # build python interface of dolfin
     @run_after('install')
-    def install_python_components(self):
-        for package in self.python_components:
-            with working_dir(join_path('depends', package)):
-                setup_py('install', '--prefix={0}'.format(self.prefix))
+    def install_python_interface(self):
+        if '+python' in self.spec:
+            cd('python')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]: It's more Spack-ish to use:
```python
with working_dir('python'):
    ...
```

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    version('2018.1.0',       sha256='c5a6511693106d1cd2fc013148d0cd01cd1b99fc65dab461ca0b95851a9ea271')
+    version('2017.2.0.post0', sha256='1969a5460cb866c478df64874ce213f81cb5c893b89f991a578e258b1a64fee5')
+    version('2016.2.0',       sha256='916c94c18047e1281d08e67c515aa1d4175320723609fbbfb0fc498ab4d41bb6')
+    version('1.6.0',          sha256='418b0b89f2e46e28611476f8269d99461c57a79b8c4e676821c2a063d3ed26c2')
+
+    depends_on('python@3:', type=('build', 'run'))
+
+    depends_on('py-setuptools', type='build')
+
+    depends_on('py-numpy')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-ffc/package.py]: Shouldn't this be type=build/run or build/link/run?

**Review by adamjstewart — COMMENTED** [review]
Note that Spack has a new feature that allows you to deprecate old versions if you're concerned about supporting them: https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 6 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: agreed, will be changed in next commit

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 7 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: agreed, will be changed in next commit

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 5 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: I agree with your first point to remove the `py-`. The reason for having the `petsc4py` and `slepcy4py` as a variant in there is for later use in the `cmake_args` function (see `self.define_from_variant` functions). Is there another and maybe more elegant way to omit the variants definitions at the beginning and to set the `cmake_args`?

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py (hunk 8 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-ffc/package.py]: Cross checking with the `py-fenics-ffcx` package I agree this should be of type `(build, run)`. I'm not sure if it is required as it currently works on my machine, however I will align it in the next commit.

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 5 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/fenics/package.py]: There's nothing wrong with the use of variants and `self.define_from_variant`, your current approach is correct.

Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: Alright, thanks Adam. But do you know if there is a way to make the `petsc4py` and `slepc4py` variants dependent on the `python` and `petsc` / `slepc` variants to avoid compatibility conflicts?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/fenics/package.py]:
```python
conflicts('+petsc4py', when='~python')
conflicts('+petsc4py', when='~petsc')
conflicts('+slepc4py', when='~python')
conflicts('+slepc4py', when='~slepc')
```

**Comment by adamjstewart** [comment]
Maintainer ping: @js947 @chrisrichardson

@emai-imcs would you like to be a maintainer of any of these packages? We should add maintainers to the new packages.

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 9]
```diff
@@ -10,26 +10,44 @@ class Fenics(CMakePackage):
...
-    the code generation interface UFC, the form language UFL and a
-    range of additional components."""
+    finite element tabulator FIAT, the just-in-time compiler Instant / Dijitso,
+    the code generation interface UFC, the form language UFL and a range of
+    additional components."""
 
     homepage = "http://fenicsproject.org/"
+    git      = "https://bitbucket.org/fenics-project/dolfin.git"
     url      = "https://bitbucket.org/fenics-project/dolfin/downloads/dolfin-1.6.0.tar.gz"
     base_url = "https://bitbucket.org/fenics-project/{pkg}/downloads/{pkg}-{version}.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]: It seems like this variable is no longer used?

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 10]
```diff
@@ -43,135 +61,105 @@ class Fenics(CMakePackage):
...
     depends_on('suite-sparse', when='+suite-sparse')
     depends_on('qt', when='+qt')
 
-    depends_on('py-ply', type=('build', 'run'))
-    depends_on('py-six', type=('build', 'run'))
-    depends_on('py-numpy', type=('build', 'run'))
-    depends_on('py-sympy', type=('build', 'run'))
-    depends_on('swig@3.0.3:', type=('build', 'run'))
-    depends_on('cmake@2.8.12:', type='build')
+    depends_on('py-pybind11', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]:
```suggestion
    depends_on('py-pybind11@2.2.4', type=('build', 'run'))
```
At least in the latest release, you'll have to see if older versions have different requirements.

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 11]
```diff
@@ -43,135 +61,105 @@ class Fenics(CMakePackage):
...
+    depends_on('cmake@3.17.3:', type='build')
 
-    depends_on('py-setuptools', type='build')
+    depends_on('py-setuptools', type='build', when='+python')
+    depends_on('py-pkgconfig', type='run', when='+python')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]:
```suggestion
    depends_on('py-pkgconfig', type=('build', 'run'), when='+python')
```

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 12]
```diff
@@ -43,135 +61,105 @@ class Fenics(CMakePackage):
...
+            depends_on('py-fenics-instant{0}'.format(wver), type=('build', 'run'), when=wver)
+        else:
+            depends_on('py-fenics-dijitso{0}'.format(wver), type=('build', 'run'), when=wver)
+        depends_on('py-fenics-ufl{0}'.format(wver), type=('build', 'run'), when=wver)
+        if ver in ['2019.1.0', '2017.2.0']:
+            wver = wver + '.post0'
+        depends_on('py-fenics-ffc{0}'.format(wver), type=('build', 'run'), when=wver)
+
+    # package dependencies
+    depends_on('python@3:', type=('build', 'run'), when='+python')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]:
```suggestion
    depends_on('python@3.5:', type=('build', 'run'), when='+python')
```

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 13]
```diff
@@ -43,135 +61,105 @@ class Fenics(CMakePackage):
...
+        wver = '@' + ver
+        depends_on('py-fenics-fiat{0}'.format(wver), type=('build', 'run'), when=wver)
+        if(Version(ver) < Version('2018.1.0')):
+            depends_on('py-fenics-instant{0}'.format(wver), type=('build', 'run'), when=wver)
+        else:
+            depends_on('py-fenics-dijitso{0}'.format(wver), type=('build', 'run'), when=wver)
+        depends_on('py-fenics-ufl{0}'.format(wver), type=('build', 'run'), when=wver)
+        if ver in ['2019.1.0', '2017.2.0']:
+            wver = wver + '.post0'
+        depends_on('py-fenics-ffc{0}'.format(wver), type=('build', 'run'), when=wver)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]: Are these only needed when `+python`?

File: var/spack/repos/builtin/packages/py-fenics-dijitso/package.py [hunk 14]
```diff
@@ -15,11 +15,11 @@ class PyFenicsDijitso(PythonPackage):
...
     version("master", branch="master")
-    version("2018.1.0", sha256="2084ada1e7bd6ecec0999b15a17db98c72e26f1ccbf3fcbe240b1a035a1a2e64")
-    version("2017.2.0", sha256="05a893d17f8a50067d303b232e41592dce2840d6499677ad8b457ba58a679b58")
-    version("2017.1.0", sha256="ef23952539d349fbd384d41302abc94413d485ca7e8c29e8a0debc2aadaf8657")
-    version("2016.2.0", sha256="1bfa0ac0d47dae75bbde8fabb1145d3caa2e6c28a1e4097ad5550a91a8a205e4")
+    version('2019.1.0', sha256='eaa45eec4457f3f865d72a926b7cba86df089410e78de04cd89b15bb405e8fd9')
+    version('2018.1.0', sha256='2084ada1e7bd6ecec0999b15a17db98c72e26f1ccbf3fcbe240b1a035a1a2e64')
+    version('2017.2.0', sha256='05a893d17f8a50067d303b232e41592dce2840d6499677ad8b457ba58a679b58')
+    version('2017.1.0', sha256='ef23952539d349fbd384d41302abc94413d485ca7e8c29e8a0debc2aadaf8657')
+    version('2016.2.0', sha256='7358c6cbeacaada3ef3071bdcf2a5f342cb7ee1ed736742d2b747b1fd4454d4c')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-dijitso/package.py]: Why did this checksum change?

File: var/spack/repos/builtin/packages/py-fenics-dijitso/package.py [hunk 15]
```diff
@@ -15,11 +15,11 @@ class PyFenicsDijitso(PythonPackage):
...
 
     depends_on("py-setuptools", type="build")
-    depends_on("python@3.5:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-dijitso/package.py]: Please re-add this line

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    git      = "https://bitbucket.org/fenics-project/ffc.git"
+    url      = "https://bitbucket.org/fenics-project/ffc/downloads/ffc-2019.1.0.post0.tar.gz"
+
+    version('2019.1.0.post0', sha256='306e1179630200a34202975a5369194939b3482eebfc34bc44ad74dab1f109e8')
+    version('2018.1.0',       sha256='c5a6511693106d1cd2fc013148d0cd01cd1b99fc65dab461ca0b95851a9ea271')
+    version('2017.2.0.post0', sha256='1969a5460cb866c478df64874ce213f81cb5c893b89f991a578e258b1a64fee5')
+    version('2016.2.0',       sha256='916c94c18047e1281d08e67c515aa1d4175320723609fbbfb0fc498ab4d41bb6')
+    version('1.6.0',          sha256='418b0b89f2e46e28611476f8269d99461c57a79b8c4e676821c2a063d3ed26c2')
+
+    depends_on('python@3:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-ffc/package.py]:
```suggestion
    depends_on('python@3.5:', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+    depends_on('py-setuptools', type=('build', 'run'))
+    depends_on('py-numpy', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-ffc/package.py]: Missing requirements:
```python
RESTRICT_REQUIREMENTS = ">=2019.1.0,<2019.2"                                             
                                                                                         
if on_rtd:                                                                               
    REQUIREMENTS = []                                                                    
else:                                                                                    
    REQUIREMENTS = [                                                                     
        "numpy",                                                                         
        "fenics-fiat%s" % RESTRICT_REQUIREMENTS,                                         
        "fenics-ufl%s" % RESTRICT_REQUIREMENTS,                                          
        "fenics-dijitso%s" % RESTRICT_REQUIREMENTS,                                      
    ]       
```

File: var/spack/repos/builtin/packages/py-fenics-fiat/package.py [hunk 18]
```diff
@@ -17,18 +17,20 @@ class PyFenicsFiat(PythonPackage):
     nonconforming elements"""
 
     homepage = "https://fenicsproject.org/"
-    url = "https://github.com/FEniCS/fiat/archive/2019.1.0.tar.gz"
-    git = "https://github.com/FEniCS/fiat.git"
+    url = "https://bitbucket.org/fenics-project/fiat/downloads/fiat-2019.1.0.tar.gz"
+    git = "https://bitbucket.org/fenics-project/fiat.git"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-fiat/package.py]: According to https://bitbucket.org/fenics-project/fiat/src/master/, the project has moved to GitHub, so the previous url/git should be used instead.

File: var/spack/repos/builtin/packages/py-fenics-instant/package.py [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    url      = "https://bitbucket.org/fenics-project/instant/downloads/instant-2017.2.0.tar.gz"
+
+    version('2017.2.0', sha256='be24f162fd1a89b82fae002db8df0b4f111fd50db83d78c0c121015c02e45b7b')
+    version('2016.2.0', sha256='2fb4f6cf087369ca11482a945ae695066f4b9aca15d90e033f059ea1107bb874')
+    version('1.6.0',    sha256='a9c569b3fa5ce5f00465323cf53da116f4cf3d416920b5628acbdcdc15b892a7')
+
+    depends_on('py-setuptools', type="build")
+    depends_on('py-numpy', type=("build", "run"))
+    depends_on('cmake', type="run")
+    depends_on('swig', type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-instant/package.py]: Missing dependencies:
```python
depends_on('python@2.7:', type=('build', 'run'))
depends_on('py-six', type=('build', 'run'))
depends_on('py-subprocess32', when='^python@:2.999', type=('build', 'run'))
```
```

File: var/spack/repos/builtin/packages/py-fenics-ufl/package.py [hunk 20]
```diff
@@ -14,17 +14,17 @@ class PyFenicsUfl(PythonPackage):
     mathematical notation."""
 
     homepage = "https://fenicsproject.org/"
-    url = "https://github.com/FEniCS/ufl/archive/2019.1.0.tar.gz"
-    git = "https://github.com/FEniCS/ufl.git"
+    url = "https://bitbucket.org/fenics-project/ufl/downloads/ufl-2019.1.0.tar.gz"
+    git = "https://bitbucket.org/fenics-project/ufl.git"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-ufl/package.py]: According to https://bitbucket.org/fenics-project/ufl/src/master/, the package has moved to GitHub, so the previous url/git should be used

File: var/spack/repos/builtin/packages/py-fenics-ufl/package.py [hunk 21]
```diff
@@ -14,17 +14,17 @@ class PyFenicsUfl(PythonPackage):
...
-    version('2017.2.0',       sha256='0adff7a511185b20c38ddaccdeed6c1b2ecafe4b163c688bfd1316d5c3b1c00d')
-    version('2017.1.0.post1', sha256='82c8170f44c2392c7e60aa86495df22cc209af50735af8115dc35aeda4b0ca96')
+    version('2019.1.0', sha256='213a56bfa2edb0365909fa6f110f970e582b10a3f5f1fd9a0e85254f1eefaa11')
+    version('2018.1.0', sha256='ed191f5e0abfeb9922a68a6802f819e9cc6c50749472c01365319f66a5cfcbea')
+    version('2017.2.0', sha256='45562fcd29ceb2f2e37c44bc2552b8e6c5d76409f73526096d2bcdb7078cf99c')
+    version('2016.2.0', sha256='1689223696e5fa4534a9db7a34bc33782fae9f5ae00d67d834d2ef16be6b1300')
+    version('1.6.0',    sha256='4296997a6a940fe02d6543c3ff24c55daa03986de95110df5a270622c718c756')
 
-    depends_on("python@3.5:", type=('build', 'run'))
+    depends_on("python@3:", type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-ufl/package.py]: Undo this change

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by adamjstewart** [comment]
P.S. It would be helpful for me if you could split this PR into separate PRs for each package.

File: var/spack/repos/builtin/packages/py-fenics-dijitso/package.py (hunk 14 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-dijitso/package.py]: The checksums were updated based on the result from the `spack checksum` command, why they have changed, I can't tell.

File: var/spack/repos/builtin/packages/py-fenics-dijitso/package.py (hunk 15 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-dijitso/package.py]: Ok, I will reintroduce it and create a new PR for the `py-fenics-dijitso` package.

File: var/spack/repos/builtin/packages/py-fenics-fiat/package.py (hunk 18 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-fiat/package.py]: The problem with the GitHub Repo was, that it only includes the FIAT version starting from 2017.1.0, when I run the 'spack checksum' command. However to add the 2016.2.0 version (for consistency with the other packages), I suggest to use the tag from the fiat git Repo.
Note that this change is outsourced to a separate PR as well.

File: var/spack/repos/builtin/packages/py-fenics-ufl/package.py (hunk 20 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-ufl/package.py]: Like with the FIAT package the UFL repository on GitHub has no .tar.gz archives, however I suggest to use the tag flag instead. As requested I will raise a separate PR.

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py (hunk 16 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-ffc/package.py]: ok

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py (hunk 17 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-ffc/package.py]: Thanks for the check, I will add the dependencies and remove them from the `fenics` package as they are implicitly covered by the `py-fenics-ffc` package. However, I don't think that the `py-fenics-dijitso` package is necessarily required for ffc. Trying to install the ffc package, it worked without `py-fenics-dijitso`.

As suggest I will raise a separate PR for the new ffc package.

File: var/spack/repos/builtin/packages/py-fenics-ufl/package.py (hunk 21 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-ufl/package.py]: done

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py (hunk 17 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-ffc/package.py]: `py-fenics-dijitso` is still required, if it is not found setuptools will try to download it, which won't work on air-gapped networks.

File: var/spack/repos/builtin/packages/py-fenics-instant/package.py (hunk 19 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-instant/package.py]: ok, added dependencies and created a new PR #21775

File: var/spack/repos/builtin/packages/py-fenics-ffc/package.py (hunk 17 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/py-fenics-ffc/package.py]: ok, `py-fenics-dijitso` package added to PR for the FFC package #21862

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 9 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: yes, `base_url` can be removed

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 11 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: agreed, suggestion will be added

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 12 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: agreed, will be changed as suggested

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 13 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: Checking the [CMakeLists.txt](https://bitbucket.org/fenics-project/dolfin/src/2019.1.0.post0/CMakeLists.txt)  for `dolfin`, the python modules FFC, FIAT, Dijitso (Instant) and UFL are only required for `dolfin` when build with the python interface. 

Can I add the `+python` to the version dependency in `when=`?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/fenics/package.py]: Yes, it would look like `when='@1.2.3: +python'`

File: var/spack/repos/builtin/packages/fenics/package.py (hunk 10 above)
Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: Checking the `dolfin` 2018 and 2019 versions, they seem to work with `py-pybind11@2.2.4`, hence I will amend the spec as suggested.

**Comment by emai-imcs** [comment]
@adamjstewart, did you found some time to review the PR? With the latest commit it should identify the `fenics`package and the two patch files as changed compared to the development branch. The `fenics` package was amended to incorporate the changes requested from the last comment cycle.

File: var/spack/repos/builtin/packages/fenics/package.py [hunk 22]
```diff
@@ -43,135 +60,105 @@ class Fenics(CMakePackage):
...
+        depends_on('py-fenics-ufl{0}'.format(wver), type=('build', 'run'), when=wver + '+python')
+        if ver in ['2019.1.0', '2017.2.0']:
+            wver = '@' + ver + '.post0'
+        depends_on('py-fenics-ffc{0}'.format(wver), type=('build', 'run'), when=wver + '+python')
+
+    # package dependencies
+    depends_on('python@3.5:', type=('build', 'run'), when='+python')
     depends_on('eigen@3.2.0:')
-    depends_on('boost+filesystem+program_options+system+iostreams+timer+regex+chrono')
+    depends_on('pkgconfig')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fenics/package.py]: I assume this should be type=build?

Comment (emai-imcs) [reply var/spack/repos/builtin/packages/fenics/package.py]: Agreed, same build type is used in `fenics-dolfinx` package. Updated `fenics` package and committed change.

**Review by adamjstewart — APPROVED** [review]

**Review by haampie — APPROVED** [review]
